src/core/thread/network_diagnostic.cpp

```diff
@@ -0,0 +1,494 @@
+            SuccessOrExit(error = message->Append(&tlv, tlv.GetSize()));
+            break;
+        }
+
+        case NetworkDiagnosticTlv::kNetworkData:
+        {
+            NetworkDataTlv tlv;
+            tlv.Init();
+            mMle.FillNetworkDataTlv((*reinterpret_cast<Mle::NetworkDataTlv *>(&tlv)), true);
+            SuccessOrExit(error = message->Append(&tlv, tlv.GetSize()));
+            break;
+        }
+
+        case NetworkDiagnosticTlv::kIPv6AddressList:
+        {
+            IPv6AddressListTlv tlv;
+            tlv.Init();
+            SuccessOrExit(error = FillIPv6AddressList(tlv));
+            SuccessOrExit(error = message->Append(&tlv, tlv.GetSize()));
+            break;
+        }
+
+        case NetworkDiagnosticTlv::kMacCounters:
+        {
+            MacCountersTlv tlv;
+            memset(&tlv, 0, sizeof(tlv));
+            tlv.Init();
+            SuccessOrExit(error = message->Append(&tlv, tlv.GetSize()));
+            break;
+        }
+
+        case NetworkDiagnosticTlv::kBatteryLevel:
+        {
+            // TODO Need more api from driver
+            BatteryLevelTlv tlv;
+            tlv.Init();
+            tlv.SetBatteryLevel(100);
+            SuccessOrExit(error = message->Append(&tlv, tlv.GetSize()));
+            break;
+        }
+
+        case NetworkDiagnosticTlv::kSupplyVoltage:
+        {
+            // TODO Need more api from driver
+            SupplyVoltageTlv tlv;
+            tlv.Init();
+            tlv.SetSupplyVoltage(0);
+            SuccessOrExit(error = message->Append(&tlv, tlv.GetSize()));
+            break;
+        }
+
+        case NetworkDiagnosticTlv::kChildTable:
+        {
+            ChildTableTlv tlv;
+            tlv.Init();
+            SuccessOrExit(error = FillChildTable(tlv));
+            SuccessOrExit(error = message->Append(&tlv, tlv.GetSize()));
+            break;
+        }
+
+        case NetworkDiagnosticTlv::kChannelPages:
+        {
+            ChannelPagesTlv tlv;
+            tlv.Init();
+            tlv.GetChannelPages()[0] = 0;
+            tlv.SetLength(1);
+            SuccessOrExit(error = message->Append(&tlv, tlv.GetSize()));
+            break;
+        }
+
+        default:
+            ExitNow();
+        }
+    }
+
+    memcpy(&messageInfo, &aMessageInfo, sizeof(messageInfo));
+    memset(&messageInfo.mSockAddr, 0, sizeof(messageInfo.mSockAddr));
+    otLogInfoNetDiag("Sending diagnostic get acknowledgment\n");
+    SuccessOrExit(error = mCoapServer.SendMessage(*message, messageInfo));
+
+    otLogInfoNetDiag("Sent diagnostic get acknowledgment\n");
+
+exit:
+
+    if (error != kThreadError_None && message != NULL)
+    {
+        message->Free();
+    }
+}
+
+void NetworkDiagnostic::HandleDiagnosticReset(void *aContext, Coap::Header &aHeader, Message &aMessage,
+                                              const Ip6::MessageInfo &aMessageInfo)
+{
+    NetworkDiagnostic *obj = reinterpret_cast<NetworkDiagnostic *>(aContext);
```

Comment:
I copied from address_resolver.cpp. I'm not clear where's the difference between the two places.